[MRG] update numpydoc to upstream, include fix for line numbers

amueller opened the pull request:
Fixes #4077.

This also includes https://github.com/numpy/numpydoc/pull/61 which fixes the gigantic long tracebacks generated by numpydoc. Now they are actually (approximately) helpful.

The new numpydoc has a Yields section which I didn't make use of yet.

It looks like having trailing underscores at attributes and not having backticks is an error again. I'm not sure why that is. I know it used to be that way. Did we hack our numpydoc at some point? Maybe @jnothman or @ogrisel or @GaelVaroquaux remember?

Oh and obviously I fixed the upstream stuff so I have less trouble fixing our docs for the release, so this would be nice to merge ;)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/7355)

<!-- Reviewable:end -->

jnothman commented:
> Did we hack our numpydoc at some point?

We certainly did. The merge with upstream is not so simple.

amueller commented:
@jnothman This PR does it. And it's not so bad. There are two changes relative to numpydoc master in here. I would argue both are bugfixes.

jnothman commented:
> This PR does it. And it's not so bad. 

Sorry I didn't read properly. Would it be worth doing a diff of the HTML generated to make sure you've not missed any differences?

amueller commented:
There are probably quite some differences, but I'll try to do a diff.

amueller commented:
There is no table for the methods with links, and the attributes are now below the notes. The attributes should be documented like the methods now. I'm not sure where the method table went, I think that was my fault.

amueller commented:
So this might indeed need a little bit more work.

amueller committed:
minor doc fixes

update numpydoc to upstream, include fix for line numbers

try to exclude numpydoc from unit test (why would we test those?!)

don't run flake8 on the sphinxext folder.

make attributes bold (and literal) so we don't have to worry about trailing underscores. https://github.com/numpy/numpydoc/pull/63

actually exclude numpydoc from tests

amueller committed:
update conf to include method table

amueller commented:
This should be good now. The new version has the Attribute section next to the Methods section, instead next to the Parameters section. Otherwise the docs should be pretty much the same.

jnothman commented:
I don't think having attributes after References makes sense in our context, even if I get why Attributes and Methods logically fit together. Should we change numpydoc to make this configurable?

jnothman commented:
Also, I think we should rename `numpy_ext` to `numpydoc`?

amueller commented:
happy to change ``numpy_ext`` to ``numpydoc``, this way I thought the diff would be easier to read (though I guess a ``git mv`` should work).

Making the order of sections configurable would be nice indeed. It's not super obvious but doable.
I'm not sure what a good order would be though. Do you think methods after references makes sense?

jnothman commented:
The ordering should emphasise the "things you should know" or "things you
might not know were there" first. It's fine to have methods after
references, as methods are mostly common (with small variation). I'd also
like to find a way to put the gallery of auto_examples higher in the file,
but am not sure how that interacts with sphinx_gallery, and didn't find a
way to do it easily when I first introduced the gallery.

On 3 November 2016 at 08:28, Andreas Mueller <notifications@github.com>
wrote:

> happy to change numpy_ext to numpydoc, this way I thought the diff would
> be easier to read (though I guess a git mv should work).
>
> Making the order of sections configurable would be nice indeed. It's not
> super obvious but doable.
> I'm not sure what a good order would be though. Do you think methods after
> references makes sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-258004679>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69nPzLj3fSxCAnYwETuVl-wdCgL-ks5q6QCLgaJpZM4J3NUn>
> .
>

jnothman commented:
I think the key point, though, is that attributes are more important for a
description of our estimators than they are in numpy docs.

On 3 November 2016 at 12:03, Joel Nothman <joel.nothman@gmail.com> wrote:

> The ordering should emphasise the "things you should know" or "things you
> might not know were there" first. It's fine to have methods after
> references, as methods are mostly common (with small variation). I'd also
> like to find a way to put the gallery of auto_examples higher in the file,
> but am not sure how that interacts with sphinx_gallery, and didn't find a
> way to do it easily when I first introduced the gallery.
>
> On 3 November 2016 at 08:28, Andreas Mueller <notifications@github.com>
> wrote:
>
>> happy to change numpy_ext to numpydoc, this way I thought the diff would
>> be easier to read (though I guess a git mv should work).
>>
>> Making the order of sections configurable would be nice indeed. It's not
>> super obvious but doable.
>> I'm not sure what a good order would be though. Do you think methods
>> after references makes sense?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-258004679>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEz69nPzLj3fSxCAnYwETuVl-wdCgL-ks5q6QCLgaJpZM4J3NUn>
>> .
>>
>
>

neerajgangwar commented:
Feature request: Is it possible to move `numpydoc` to `sklearn/externals`? It will help in #7793.

amueller commented:
yes, but we need https://github.com/numpy/numpydoc/issues/77 first. Hurray for cross-project dependency chains!

jnothman commented:
We can do this now... :)))

jnothman commented:
I would like to see this happen. It holds up some other fixes. But I'd also rather see us depend on numpydoc as a pypi dependency rather than including the source. WDYT?

amueller commented:
@jnothman sorry for absence. I'm all for declaring dependencies instead of copying the code, but I'm not sure how @GaelVaroquaux feels about it...

jnothman commented:
we have extra doc build dependencies anyway.

On 16 May 2017 7:51 am, "Andreas Mueller" <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> sorry for absence. I'm all for
> declaring dependencies instead of copying the code, but I'm not sure how
> @GaelVaroquaux <https://github.com/gaelvaroquaux> feels about it...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-301615489>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_bBmu5G8vagfPtyIFL_38VaK0t6ks5r6MjqgaJpZM4J3NUn>
> .
>

amueller commented:
True, I'll update the PR soonish

jnothman commented:
it's good to have you back. I hope your teaching went swimmingly!

On 16 May 2017 8:42 am, "Andreas Mueller" <notifications@github.com> wrote:

> True, I'll update the PR soonish
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-301625738>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_7AfHNJtSwenevtcEb-SDx4ImGGks5r6NThgaJpZM4J3NUn>
> .
>

amueller commented:
Thanks :) got good feedback so far, let's see what the course evaluations say.

amueller committed:
remove numpy_ext because I don't want to maintain it.

amueller commented:
So current numpydoc master says
```
    raise ValueError(msg)
ValueError: The section Notes appears twice in the docstring.
```
... without any context...

amueller commented:
Fix here: https://github.com/numpy/numpydoc/pull/67

amueller committed:
remove duplicate notes section

amueller commented:
Hm there are many
```
  File "/home/andy/anaconda3/lib/python3.5/site-packages/Sphinx-1.5-py3.5.egg/sphinx/ext/autodoc.py", line 547, in import_object
    __import__(self.modname)
ImportError: No module named 'sklearn.preprocessing.data.MinMaxScaler'; 'sklearn.preprocessing.data' is not a package
/home/andy/checkout/scikit-learn/doc/modules/generated/sklearn.preprocessing.MultiLabelBinarizer.rst:63: WARNING: autodoc: failed to import method 'MultiLabelBinarizer.__init__' from module 'sklearn.preprocessing.label.MultiLabelBinarizer'; the following exception was raised:
Traceback (most recent call last):
```
I have no idea where that comes from...

amueller commented:
Also docs don't build with current sphinx because https://github.com/sphinx-gallery/sphinx-gallery/pull/242

amueller commented:
And the autodoc warning is still there....

amueller commented:
[Ok, that's enough debugging sphinx-gallery, numpydoc and sphinx for a day]

jnothman commented:
wow. I'll try give it a second pair of eyes, but I've got teaching still!

On 19 May 2017 6:53 am, "Andreas Mueller" <notifications@github.com> wrote:

> [Ok, that's enough debugging sphinx-gallery, numpydoc and sphinx for a day]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-302538190>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_39Z4G2QPSfs5TgxIxnaIo6bYahks5r7K_WgaJpZM4J3NUn>
> .
>

lesteve commented:
If we go for `numpy_ext` as an external dependency, it needs to be installed on CircleCI.

amueller committed:
install numpydoc on circleci

amueller committed:
sphinx_gallery hotfix

amueller committed:
Merge branch 'master' into numpydoc_upstream

# Conflicts:
#	doc/sphinxext/sphinx_gallery/gen_gallery.py

amueller committed:
add numpydoc template for classes

amueller commented:
Current changes wrt master: ``__init__`` is not documented:
master:

![image](https://user-images.githubusercontent.com/449558/26876635-684c19a8-4b87-11e7-9dc8-027e98332e42.png)

this:
![image](https://user-images.githubusercontent.com/449558/26876654-818125f8-4b87-11e7-84ca-511aacfcf29b.png)

This is redundant with the class definition at the top of the page, so not an issue imho.

As you can also see in this screenshots, the method names are not links. I'm not sure why that is, and that seems like an issue to me.

finally, the attributes are now formatted differently:
master:
![image](https://user-images.githubusercontent.com/449558/26876704-b59d0d02-4b87-11e7-945f-edf33325f5de.png)


this:
![image](https://user-images.githubusercontent.com/449558/26876686-a831e764-4b87-11e7-8ed6-22b680ea0e39.png)

This can not be changed without adding additional switches to numpydoc master, and I don't think it's big enough an issue to block this PR.

amueller commented:
The missing links happen not always, and think it's related to the import errors that are produced, see https://github.com/sphinx-doc/sphinx/issues/3768

amueller committed:
fix typ in template file

amueller committed:
remove old misnamed file

jnothman reviewed (commented):
It's annoying but I think we should *not* make this a release priority.

amueller commented:
This was fixed on numpydoc master. If we want to rely on numpydoc release, we would need to wait for them to make a release - which might or might not happen before ours. Not a priority, but if we can agree that the changes are ok, the PR should be simple to review and we can decide with short notice whether to include or not? I'll certainly be using only this branch to fix sphinx errors in master, because anything else is insane.

amueller commented:
This was fixed on numpydoc master. If we want to rely on numpydoc release, we would need to wait for them to make a release - which might or might not happen before ours. Not a priority, but if we can agree that the changes are ok, the PR should be simple to review and we can decide with short notice whether to include or not? I'll certainly be using only this branch to fix sphinx errors in master, because anything else is insane.

amueller committed:
Merge branch 'master' into numpydoc_upstream

# Conflicts:
#	sklearn/datasets/species_distributions.py

amueller committed:
undo skipping removed numpy_ext folder

amueller commented:
right now the circle build doesn't have the template mechanism for moving the attributes up. But otherwise it looks good. We can either wait for a numpydoc release (which seems imminent) or pip install the github version. Without this it's near impossible to fix doc build errors....

jnothman commented:
I know we need to merge this, but i don't see it as release critical.

On 20 Jun 2017 1:37 am, "Andreas Mueller" <notifications@github.com> wrote:

> right now the circle build doesn't have the template mechanism for moving
> the attributes up. But otherwise it looks good. We can either wait for a
> numpydoc release (which seems imminent) or pip install the github version.
> Without this it's near impossible to fix doc build errors....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-309478284>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65J1fMNnqFDbHZhaU-M7IWT4MdpPks5sFpWfgaJpZM4J3NUn>
> .
>

amueller commented:
yeah, it's not really ;)

amueller commented:
The thing is more that it helps get rid of all the issues in the website. Cherry picking to and fro from this branch is pretty annoying.

amueller commented:
Numpydoc was just released, I'm triggering Circle so we can see what it looks like.

ogrisel commented:
The files `doc/sphinxext/README.txt` and `doc/sphinxext/LICENSE.txt` should be deleted (I think).

amueller committed:
Merge branch 'master' into numpydoc_upstream

# Conflicts:
#	sklearn/datasets/species_distributions.py

amueller committed:
removing numpydoc license and readme

amueller commented:
@ogrisel done :)

jnothman commented:
This has drastically changed the attributes listing. See any class, e.g. http://scikit-learn.org/dev/modules/generated/sklearn.cluster.bicluster.SpectralBiclustering.html

jnothman commented:
Compare http://scikit-learn.org/0.19/modules/generated/sklearn.cluster.bicluster.SpectralBiclustering.html

jnothman commented:
Numpydoc changed Attributes from a Parameter List to a Member List.

jnothman reviewed (commented):
Attributes issue raised at https://github.com/numpy/numpydoc/issues/102

amueller commented:
@jnothman I though we were fine with that? I posted screenshots above.

jnothman commented:
I didn't get around to reviewing it. It's unreadable IMO, and the attribute
documentation is pretty essential in scikit-learn.

On 22 July 2017 at 04:44, Andreas Mueller <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> I though we were fine with that?
> I posted screenshots above.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-317081750>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6z_ezSxkn5x_klPjXiBibtEKD-Rcks5sQPGbgaJpZM4J3NUn>
> .
>

amueller commented:
I don't think it's unreadable, but yeah, it was better before. So we could revert, monkey patch, or fix in upstream and use a dev version to build the website?

jnothman commented:
I'm trying to work out something with numpydoc. it's okay to leave it this
way in dev, for now. definitely don't pull this into 0.19.x though

On 23 Jul 2017 11:55 am, "Andreas Mueller" <notifications@github.com> wrote:

> I don't think it's unreadable, but yeah, it was better before. So we could
> revert, monkey patch, or fix in upstream and use a dev version to build the
> website?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-317222818>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6y011h0gtRhSaw2ZndaVa_CSPVakks5sQqfzgaJpZM4J3NUn>
> .
>

jnothman commented:
A particularly bad case of attributes at http://scikit-learn.org/dev/modules/generated/sklearn.model_selection.GridSearchCV.html: https://github.com/numpy/numpydoc/issues/104

jnothman commented:
A particularly bad case of attributes at http://scikit-learn.org/dev/modules/generated/sklearn.model_selection.GridSearchCV.html. See https://github.com/numpy/numpydoc/issues/104. Perhaps we should consider reverting before we get bug reports about this, pending some fixes upstream that I hope I find time to make.

amueller commented:
That's indeed pretty bad. I'm slightly against reverting because it makes error messages unreadable. But I guess readable docs is more important than readable doc-building error messages.

jnothman commented:
I'll have a fix for that one very soon i expect

On 25 Jul 2017 3:10 am, "Andreas Mueller" <notifications@github.com> wrote:

That's indeed pretty bad. I'm slightly against reverting because it makes
error messages unreadable. But I guess readable docs is more important than
readable doc-building error messages.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/scikit-learn/scikit-learn/pull/7355#issuecomment-317489902>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAEz63ndxnK87hGtmsRhHrbX3hcMilIYks5sRNABgaJpZM4J3NUn>
.
